### PR TITLE
Simplifying update sources, new keybase.io (api) source (v2)

### PR DESCRIPTION
**Pull request by gabriel**
This is the update source that makes request to keybase.io

Removes the sources prerelease tag which was confusing (and currently always on).

### CODE REVIEW

**Comment by gabriel** [comment]
@keybase/updater-hackers

File: keybase/source.go [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+// UpdateSource finds releases/updates on keybase.io
+type UpdateSource struct {
+	log      logging.Logger
+	endpoint string
+}
+
+// NewUpdateSource contructs an update source for keybase.io
+func NewUpdateSource(log logging.Logger) UpdateSource {
+	return newUpdateSource("https://keybase.io/_/api/1.0/pkg/update.json", log)
```
Comment (chrisnojima) [inline keybase/source.go]: should this be api.keybase.io?

File: keybase/source.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+		return nil, err
+	}
+	urlValues := url.Values{}
+	urlValues.Add("install_id", options.InstallID)
+	urlValues.Add("version", options.Version)
+	urlValues.Add("platform", options.Platform)
+	urlValues.Add("run_mode", options.Env)
+	urlValues.Add("os_version", options.OSVersion)
+	urlValues.Add("upd_version", options.UpdaterVersion)
+	u.RawQuery = urlValues.Encode()
```
Comment (chrisnojima) [inline keybase/source.go]: does this err out?

**Comment by chrisnojima** [comment]
@jzila

File: keybase/source.go [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+	urlString := u.String()
+
+	req, err := http.NewRequest("GET", urlString, nil)
+	client := &http.Client{
+		Timeout: time.Minute,
+	}
+	k.log.Infof("Request %#v", urlString)
+	resp, err := client.Do(req)
+	defer func() { _ = util.DiscardAndCloseBody(resp) }()
```
Comment (jzila) [inline keybase/source.go]: why does this need to be in a separate closure instead of `defer util.DiscardAndCloseBody(resp)`?

File: remote.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+// FindUpdate returns update for options
+func (r RemoteUpdateSource) FindUpdate(options UpdateOptions) (update *Update, err error) {
+	sourceURL := r.sourceURL(options)
+	req, err := http.NewRequest("GET", sourceURL, nil)
+	client := &http.Client{
+		Timeout: time.Minute,
+	}
+	r.log.Infof("Request %#v", sourceURL)
+	resp, err := client.Do(req)
+	defer func() { _ = util.DiscardAndCloseBody(resp) }()
```
Comment (jzila) [inline remote.go]: here too

File: keybase/source.go [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+
+	var reader io.Reader = resp.Body
+	var res updateResponse
+	if err = json.NewDecoder(reader).Decode(&res); err != nil {
+		err = fmt.Errorf("Invalid API response %s", err)
+		return
+	}
+
+	if res.Status.Code != 0 {
+		err = fmt.Errorf("API returned error response: %#v", res)
```
Comment (jzila) [inline keybase/source.go]: you can return here and avoid the else

File: local.go [hunk 6]
```diff
@@ -35,7 +38,7 @@ func digest(URL string) (digest string, err error) {
 	if err != nil {
 		return
 	}
-	defer f.Close()
+	defer func() { _ = f.Close() }()
```
Comment (jzila) [inline local.go]: why put this into a new function?

File: local.go [hunk 7]
```diff
@@ -50,16 +53,19 @@ func readFile(path string) (string, error) {
 	if err != nil {
 		return "", err
 	}
-	defer sigFile.Close()
+	defer func() { _ = sigFile.Close() }()
```
Comment (jzila) [inline local.go]: why put this into a new function?

File: remote.go [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	}
+}
+
+// Description returns update source description
+func (r RemoteUpdateSource) Description() string {
+	return "Remote"
+}
+
+func (r RemoteUpdateSource) sourceURL(options UpdateOptions) string {
+	params := util.JoinPredicate([]string{options.Platform, options.Env, options.Channel}, "-", func(s string) bool { return s != "" })
```
Comment (jzila) [inline remote.go]: can't this result in an ambiguous URI? `[]string{"foo", "", "foo"}` will resolve to the same URI as `[]string{"", "foo", "foo"}`.

File: remote.go [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+		return
+	}
+
+	var reader io.Reader = resp.Body
+	var obj Update
+	if err = json.NewDecoder(reader).Decode(&obj); err != nil {
+		err = fmt.Errorf("Bad updater remote response %s", err)
+		return
+	}
+	update = &obj
```
Comment (jzila) [inline remote.go]: probably cleaner just to do:
```go
r.log.Debugf(...)
return &obj, nil
```

**Comment by jzila** [comment]
Finished first pass

File: local.go (hunk 6 above)
Comment (gabriel) [inline local.go]: gometalinter! This code passes all go linters possible, one of which checks that err is not handled

File: remote.go (hunk 8 above)
Comment (gabriel) [inline remote.go]: yeah that's ok though

File: keybase/source.go (hunk 1 above)
Comment (gabriel) [inline keybase/source.go]: That requires adding our cert... created story for that here: https://keybase.atlassian.net/browse/DESKTOP-925

File: local.go (hunk 6 above)
Comment (jzila) [inline local.go]: -_- seems a little ridiculous to create a closure to satisfy a linter...

File: keybase/source.go (hunk 3 above)
Comment (gabriel) [inline keybase/source.go]: See gometalinter comment

File: remote.go (hunk 4 above)
Comment (gabriel) [inline remote.go]: See gometalinter comment

File: local.go (hunk 7 above)
Comment (gabriel) [inline local.go]: See gometalinter comment

File: remote.go (hunk 9 above)
Comment (gabriel) [inline remote.go]: Yeah cleaned this up a bit

File: keybase/source.go (hunk 5 above)
Comment (gabriel) [inline keybase/source.go]: Yeah cleaned this up a bit

**Comment by jzila** [comment]
I don't love the closure, but if it makes you happy to pass the metalinter then I'll allow it. :shipit:

**Comment by gabriel** [comment]
Yeah those are weird... I added util methods with no error return value defined and then fixed those defer calls

File: local.go (hunk 6 above)
Comment (gabriel) [inline local.go]: The errcheck linter is a really good one to satisfy though :)

But yeah, adding util methods that don't return error and using those is a better way to satisfy the errcheck and say, I don't care if it errors.

**Pull request merged**